PHP 8.1 support

MatyCZ opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Initial PHP 8.1 support.

Fixes #8

Commit:
PHP 8.1 support

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
PHP 8.1 support

Commit:
PHP 8.1 support

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
PHP 8.1 support

Ocramius commented on src/DbSelect.php:
```diff
@@ -82,7 +84,7 @@ public function __construct(
      * @return array
      */
-    public function getItems($offset, $itemCountPerPage)
+    public function getItems($offset, $itemCountPerPage): array
```
Adding type declarations here is a problem, because this class is not `final`, so subtypes are potentially broken.

MatyCZ replied:
Removed in new PR

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -21,7 +23,7 @@ public function __invoke()
      * @return array
      */
-    public function getPaginatorConfig()
+    public function getPaginatorConfig(): array
```
Adding type declarations here is a problem, because this class is not `final`, so subtypes are potentially broken.

MatyCZ replied:
Removed in new PR

Ocramius reviewed (changes requested):
Please check CI failures too

Commit:
Remove some type hints

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
Remove some type hints

MatyCZ commented:
CI failures are caused by wrong type hints in another libraries (laminas-db), different dependencies (CI test on PHP 7.3 fails due to psr/container needs PHP 7.4, ...) or wrong return type of Mock.

Could you please check CI failures? There is probably nothing to fix.

MatyCZ commented:
I created another pull request https://github.com/laminas/laminas-db/pull/236. This will fix psalm failure because of wrong type in `Sql\Select::having()`.

tobias-trozowski commented on composer.json:
```diff
@@ -15,15 +15,15 @@
     },
     "require": {
-        "php": "^7.3 || ~8.0.0",
+        "php": "^7.3 || ~8.0.0 || ~8.1.0",
```
```suggestion
        "php": "^7.4 || ~8.0.0 || ~8.1.0",
```
IMO 7.3 support can be droppeed here too as many (most?) other laminas packages already dropped it.

tobias-trozowski commented on composer.json:
```diff
@@ -15,15 +15,15 @@
     "require-dev": {
-        "laminas/laminas-coding-standard": "~2.1.4",
+        "laminas/laminas-coding-standard": "~2.3.0",
         "phpunit/phpunit": "^9.3",
```
```suggestion
        "phpunit/phpunit": "^9.5.8",
```
bumping phpunit would make it compatible in all combinations (7.4, 8.0, 8.1, lowest, locked, latest)

Commit:
Drop PHP 7.3 support. Add property type hints.

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
Drop PHP 7.3 support. Add property type hints.

froschdesign commented on src/DbSelect.php:
```diff
@@ -103,7 +105,7 @@ public function getItems($offset, $itemCountPerPage)
      * @throws MissingRowCountColumnException
      */
-    public function count()
+    public function count(): int
```
Please remove all changes like this because the classes are not final and this means a BC break.
Thanks in advance! 👍

froschdesign commented on src/DbSelect.php:
```diff
@@ -123,7 +125,7 @@ public function count()
      * @return Select
      */
-    protected function getSelectCount()
+    protected function getSelectCount(): Select
```
The same here.

froschdesign commented on src/DbSelect.php:
```diff
@@ -150,7 +152,7 @@ protected function getSelectCount()
      * @return array
      */
-    public function getArrayCopy()
+    public function getArrayCopy(): array
```
The same here.

froschdesign commented on src/DbSelect.php:
```diff
@@ -161,10 +163,11 @@ public function getArrayCopy()
     }
 
     /**
+     * @param  array $row
      * @return int
```
These two lines can be removed because the method signature already gives the same information.

Commit:
Remove return type hints on public methods

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
Remove return type hints on public methods

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -9,7 +11,7 @@ class ConfigProvider
      * @return array
      */
-    public function __invoke()
+    public function __invoke(): array
```
This class is not `final`, so we can't add return type declarations, as they may break subtypes

Commit:
Remove return type hints on public methods

Commit:
Remove return type hints on public methods

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Ocramius commented:
A `composer update` with lowest dependencies (executed on a PHP 7.4 environment) is needed here

Commit:
Executed composer update on a PHP 7.4

Commit:
Executed composer update on a PHP 7.4

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

froschdesign commented:
@MatyCZ 
Please check the results of the code sniffer:

```
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$sql has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$select has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$countSelect has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$resultSetPrototype has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$rowCount has useless @var annotation.
```
https://github.com/laminas/laminas-paginator-adapter-laminasdb/pull/9/files#diff-f8dc66f5df7573cebf9d8d594235ad4b1510d2fe3372bf8dd8a05bde4756129dR25

And the Psalm error:
```
Error: src/DbTableGateway.php:41:29: PossiblyInvalidArgument: Argument 1 of Laminas\Db\Sql\Select::having expects Closure|Laminas\Db\Sql\Where|array<array-key, mixed>|string, possibly different type Closure|Laminas\Db\Sql\Having|non-empty-array<array-key, mixed>|non-falsy-string provided (see https://psalm.dev/092)
```
https://github.com/laminas/laminas-paginator-adapter-laminasdb/runs/4583362117?check_suite_focus=true#step:3:289

Thanks in advance! 👍

Commit:
Fix code sniffer errors

Commit:
Fix code sniffer errors

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

MatyCZ commented:
@froschdesign   I am unable to fix psalm error:
https://github.com/laminas/laminas-paginator-adapter-laminasdb/pull/9#issuecomment-996175983

Commit:
Removed composer allow-plugins config

Commit:
Removed composer allow-plugins config

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

weierophinney committed:
qa: mark phpcs and package versions plugins as allowed

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: add PossiblyInvalidArgument to Psalm baseline

Upstream library has an incorrect annotation, which means the correctly marked one here is being incorrectly flagged as invalid.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@MatyCZ I pushed an update to the Psalm baseline to ignore that error; it's a problem with the upstream laminas-db library, so we can ignore it for now.

However... we cannot accept this patch without signoff. You can add it to your existing commits by following [the directions in our contributor document](https://github.com/laminas/.github/blob/main/CONTRIBUTING.md#adding-signoffs-after-the-fact). Once you've done that, ping me, and I'll merge.

Commit:
Merge remote-tracking branch 'origin/1.1.x' into 1.1.x

MatyCZ commented:
@weierophinney ping

froschdesign commented on src/DbSelect.php:
```diff
@@ -20,32 +22,11 @@ class DbSelect implements AdapterInterface
-     * @var int
-     */
-    protected $rowCount;
+    protected Sql $sql;
```
This is a BC break because the current class is not final and the properties are _only_ protected.

MatyCZ replied:
Removed

Commit:
Remove $sql property type hint

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
Remove $sql property type hint

Commit:
Merge remote-tracking branch 'origin/1.1.x' into 1.1.x

tobias-trozowski commented on src/DbSelect.php:
```diff
@@ -22,30 +24,10 @@ class DbSelect implements AdapterInterface
-     * @var int
-     */
-    protected $rowCount;
+    protected Select $select;
```
```suggestion
    /** @var Select */
    protected $select;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

tobias-trozowski commented on src/DbSelect.php:
```diff
@@ -22,30 +24,10 @@ class DbSelect implements AdapterInterface
-     */
-    protected $rowCount;
+    protected Select $select;
+    protected ?Select $countSelect = null;
```
```suggestion
    /** @var Select|null */
    protected $countSelect;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

tobias-trozowski commented on src/DbSelect.php:
```diff
@@ -22,30 +24,10 @@ class DbSelect implements AdapterInterface
-    protected $rowCount;
+    protected Select $select;
+    protected ?Select $countSelect = null;
+    protected ResultSetInterface $resultSetPrototype;
```
```suggestion
    /** @var ResultSetInterface */
    protected $resultSetPrototype;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

tobias-trozowski commented on src/DbSelect.php:
```diff
@@ -22,30 +24,10 @@ class DbSelect implements AdapterInterface
+    protected Select $select;
+    protected ?Select $countSelect = null;
+    protected ResultSetInterface $resultSetPrototype;
+    protected ?int $rowCount = null;
```
```suggestion
    /** @var null|int*/
    protected $rowCount;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

Commit:
Remove properties type hint

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Commit:
Remove properties type hint

Signed-off-by: Matej Szendi <matej.szendi@bcom.cz>

Ocramius reviewed (approved):
LGTM, thanks @MatyCZ!
